### DISCUSSION TITLE
Spikes in data

### DISCUSSION THREAD

**Discussion by FilipDem** [addiejanssen/domoticz-solaredge-modbustcp-plugin#56] [General]
Hi,
I am running now since a while the plugin! Thanks for this! Great job!
I bought the house since a couple of months and started since then with the Domoticz plugin.

Unfortunately from time to time (once/month?) I got a spike that is messing up the charts (see image).
At a certain moment, I receive a big negative value, being the value of the total generated energy since the installation became active (but then negative)... Then a while nothing and then the same value, but with a positive value...

Anybody reported already such a problem and is there a solution/workaround?

![image](https://user-images.githubusercontent.com/16196363/208661783-82af59d8-57cc-4ee9-98eb-97a8871a5d72.png)

**Comment by janreimen** [comment]
Hello,

I encountered the same issue on 18-20 december, but not with the domoticz-solaredge-modbustcp-plugin 2.0.2, but with the official WebAPI. 

What I saw in the exportable .json / .xml's of the SolarEdge site is that there was wrong data at the source ... 

However, I can't find an explanation about your finding with the modbus plugin ... and I did not encounter the issue till now

**Comment by janreimen** [comment]
> Hi, I am running now since a while the plugin! Thanks for this! Great job! I bought the house since a couple of months and started since then with the Domoticz plugin.
> 
> Unfortunately from time to time (once/month?) I got a spike that is messing up the charts (see image). At a certain moment, I receive a big negative value, being the value of the total generated energy since the installation became active (but then negative)... Then a while nothing and then the same value, but with a positive value...
> 
> Anybody reported already such a problem and is there a solution/workaround?
> 
> ![image](https://user-images.githubusercontent.com/16196363/208661783-82af59d8-57cc-4ee9-98eb-97a8871a5d72.png)

I might have an origin : @FilipDem do you use more than one inverter on your installation ?
- as I have a 3phase SE16K (being slave) connected to the solar panels, and between him and my Solar P1 Counter is an additional SE6000H 1phase (Master). At the moment I encountered the issue again 5 mins ago, the SolarEdge Data presented as follows : 
Sum kWh of Slave = y, last updated 13h38
Sum kWh of Master = x, last updated 13h51
normally, the Sum of both should be the values of x+y, but in the scenario you describe and I encouter, the sum is y !

This could explain the negative drop followed by the regain of all produced energy (in my case).

Bypassable via dzVents script replacing a native Device ... :)

**Comment by FilipDem** [comment]
@janreimen : No I only have one inverter on my installation...
Any idea how we could investigate post-problem?
Is my assumption correct that the inverter probably returns 0 (zero) for the total_energy value? Then I think we/I can make a workaround...

**Comment by janreimen** [comment]
@FilipDem  : I don't have the exact official answer from SolarEdge, but when I look at my realtime data / logs in my installation, I tend to agree with your assumption. 

Until now (I'm using domoticz since 2 years), the issue never happened for me at the end-of-day, so the wrong data was only on the Meter/Multimeter Table and not in the Meter_Calendar / Multimeter_Calendar table. Even though, the Web Interface allows in this table to delete on unique point (i.ex. on a Temperature Node, you can delete ex-post a wrong value like 0° on the Modbus Temperature of the Inverter ... ). Alternatively, you pass as DBA to the domoticz.db file with a CLI like sqlite3 or a Frontend like sqlitebrowser ...

**Comment by addiejanssen** [comment]
I have moved this topic from "issues" to "discussions".

**Comment by FilipDem** [comment]
In the meantime I got again the problem and was able to capture the debug data. You can view the comparison between good readings and wrong modbus readings in the table below... In case wrong readings, the absolute energy_total has value 0xFFFFFFFF and several fields have absolute values of 0x8000.

I propose a patch to detect the wrong values and skip processing (see code extract below). I have this running at my side now.
What is your opinion about this patch?

```
...
if "Mode5" in Parameters and (Parameters["Mode5"] == "Extra" or Parameters["Mode5"] == "Debug"):
    to_log = inverter_values
    if "c_serialnumber" in to_log:
        to_log.pop("c_serialnumber")
    Domoticz.Log("inverter values: {}".format(json.dumps(to_log, indent=4, sort_keys=False)))

****** CODE ADDED
# Check inverter_values to avoid wrong values/spikes - occurs regularly that the following wrong values are returned
#   FFFFFFFFhex/4294967295dec: energy_total
# FFFFFF8000hex/    -32768dec: voltage_scale, power_ac, power_ac_scale, frequency_scale, power_apparent, power_apparent_scale, power_reactive, power_reactive_scale, power_factor, power_factor_scale, energy_total_scale, current_dc_scale, voltage_dc_scale, power_dc, power_dc_scale, temperature, temperature_scale

wrong_values_count = sum(value == '4294967295' or value == '-32768' for value in inverter_values.values())
energy_total = int(inverter_values[next(unit[Column.MODBUSNAME] for unit in self._LOOKUP_TABLE if unit[Column.ID] == Unit.ENERGY_TOTAL)])
Domoticz.Debug('--> checking conditions to detect wrong values... (wrong_values_count: {} - energy_total: {})'.format(wrong_values_count, energy_total))
if energy_total == 4294967295 or wrong_values_count >= 10:
    Domoticz.Status('Wrong modbus invertor values received: {}'.format(json.dumps(to_log, indent=4, sort_keys=False)))
    return
****** CODE ADDED

# Just for cosmetics in the log

updated = 0
device_count = 0

...
``` 
![inverter_values](https://user-images.githubusercontent.com/16196363/212539177-bdde2318-a515-4832-bea6-8a700259d790.jpg)

**Reply by addiejanssen** [reply]
Hi @FilipDem,

Thanks for your feedback. Next to your spike-topic, I see other data issues that could be solved by defining "boundaries" for the values returned by the inverter. I will have a look at an approach to handle data that is out of bounds.
